lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1596 @@
+//! The top-level network map tracking logic lives here.
+
+use bitcoin::secp256k1::key::PublicKey;
+use bitcoin::secp256k1::Secp256k1;
+use bitcoin::secp256k1;
+
+use bitcoin::hashes::sha256d::Hash as Sha256dHash;
+use bitcoin::hashes::Hash;
+use bitcoin::blockdata::script::Builder;
+use bitcoin::blockdata::opcodes;
+
+use chain::chaininterface::{ChainError, ChainWatchInterface};
+use ln::features::{ChannelFeatures, NodeFeatures};
+use ln::msgs::{DecodeError,ErrorAction,LightningError,RoutingMessageHandler,NetAddress};
+use ln::msgs;
+use util::ser::{Writeable, Readable, Writer, ReadableArgs};
+use util::logger::Logger;
+
+use std::cmp;
+use std::sync::{RwLock,Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::collections::BTreeMap;
+use std::collections::btree_map::Entry as BtreeEntry;
+use std;
+
+/// Receives network updates from peers to track view of the network.
+pub struct NetGraphMsgHandler {
+	secp_ctx: Secp256k1<secp256k1::VerifyOnly>,
+	/// Representation of the payment channel network
+	pub network_graph: RwLock<NetworkGraph>,
+	chain_monitor: Arc<ChainWatchInterface>,
+	full_syncs_requested: AtomicUsize,
+	logger: Arc<Logger>,
+}
+
+impl NetGraphMsgHandler {
+	/// Creates a new tracker of the actual state of the network of channels and nodes.
+	pub fn new(chain_monitor: Arc<ChainWatchInterface>, logger: Arc<Logger>) -> Self {
+		NetGraphMsgHandler {
+			secp_ctx: Secp256k1::verification_only(),
+			network_graph: RwLock::new(NetworkGraph {
+				channels: BTreeMap::new(),
+				nodes: BTreeMap::new(),
+			}),
+			full_syncs_requested: AtomicUsize::new(0),
+			chain_monitor,
+			logger: logger.clone(),
+		}
+	}
+
+	/// Get network addresses by node id
+	pub fn get_addresses(&self, pubkey: &PublicKey) -> Option<Vec<NetAddress>> {
+		let network = self.network_graph.read().unwrap();
+		network.get_nodes().get(pubkey).map(|n| n.addresses.clone())
+	}
+
+	/// Dumps the entire network view of this NetGraphMsgHandler to the logger provided in the constructor at
+	/// level Trace
+	pub fn trace_state(&self) {
+		log_trace!(self, "{}", self.network_graph.read().unwrap());
+	}
+
```

Comment:
nit: careful with excess whitespace in a number of places.

Reply:
found two, i hope that was it..